File: retaRust/python_arch_reference/reta_architecture/console_io.py
from __future__ import annotations

"""Stage-39 activated console/help/wrapping morphisms for Reta's center facade.

Historically ``libs/center.py`` owned a mixed cluster of small runtime helpers:
help-file rendering, terminal-width discovery, debug printing, CLI output,
chunking, ordered default dictionaries and order-preserving uniqueness.  Stage
39 moves that concrete logic into the architecture package while ``center.py``
keeps the historical function names as wrappers.

Mathematical reading:

* a help Markdown file is a local documentation section;
* terminal width and hyphenator state form an output-context open set;
* CLI printing is a morphism from rendered text sections to console effects;
* ordered uniqueness and chunking are finite-section morphisms used by table and
  generated-relation code;
* the old ``center`` API and the new architecture API are connected by a
  natural transformation: both routes must produce the same visible text/data
  sections.
"""

import os
import pprint
import re
import sys
from collections import OrderedDict
from collections.abc import Callable as _CallableABC
from itertools import filterfalse
from pathlib import Path
from typing import Callable, Iterable, Iterator, Optional, Sequence

try:
    from rich.console import Console as _RichConsole
    from rich.markdown import Markdown as _RichMarkdown
    from rich.syntax import Syntax as _RichSyntax
except (ModuleNotFoundError, ImportError):
    class _RichConsole:
        def __init__(self, *args, **kwargs):
            pass

        def print(self, *args, **kwargs):
            end = kwargs.get("end", "\n")
            text = " ".join(str(arg) for arg in args)
            __import__("builtins").print(text, end=end)

    class _RichMarkdown:
        def __init__(self, text, *args, **kwargs):
            self.text = text

        def __str__(self):
            return self.text

    class _RichSyntax:
        def __init__(self, text, *args, **kwargs):
            self.text = text

        def __str__(self):
            return self.text


Console = _RichConsole
Markdown = _RichMarkdown
Syntax = _RichSyntax


class DefaultOrderedDict(OrderedDict):
    """OrderedDict with a default factory, kept byte-compatible with center.py use."""

    # Source: http://stackoverflow.com/a/6190500/562769
    def __init__(self, default_factory=None, *a, **kw):
        if default_factory is not None and not isinstance(default_factory, _CallableABC):
            raise TypeError("first argument must be callable")
        OrderedDict.__init__(self, *a, **kw)
        self.default_factory = default_factory

    def __getitem__(self, key):
        try:
            return OrderedDict.__getitem__(self, key)
        except KeyError:
            return self.__missing__(key)

    def __missing__(self, key):
        if self.default_factory is None:
            raise KeyError(key)
        self[key] = value = self.default_factory()
        return value

    def __reduce__(self):
        if self.default_factory is None:
            args = tuple()
        else:
            args = (self.default_factory,)
        return type(self), args, None, None, self.items()

    def copy(self):
        return self.__copy__()

    def __copy__(self):
        return type(self)(self.default_factory, self)

    def __deepcopy__(self, memo):
        import copy

        return type(self)(self.default_factory, copy.deepcopy(self.items()))

    def __repr__(self):
        return "OrderedDefaultDict(%s, %s)" % (
            self.default_factory,
            OrderedDict.__repr__(self),
        )


def chunks(sequence: Sequence, size: int) -> Iterator[Sequence]:
    """Yield successive ``size``-sized chunks, preserving legacy helper semantics."""
    for index in range(0, len(sequence), size):
        yield sequence[index : index + size]


def unique_everseen(iterable: Iterable, key: Optional[Callable] = None, ordered_set_factory=None):
    """Yield unique elements while preserving order.

    This mirrors the historical ``center.unique_everseen`` implementation.  The
    ordered set factory is injectable so ``center.py`` can preserve its optional
    dependency on ``orderedset.OrderedSet`` while the architecture module stays
    usable without that package.
    """
    ordered_set_factory = ordered_set_factory or set
    seen = ordered_set_factory()
    seen_add = seen.add
    if key is None:
        for element in filterfalse(seen.__contains__, iterable):
            seen_add(element)
            yield element
    else:
        for element in iterable:
            marker = key(element)
            if marker not in seen:
                seen_add(marker)
                yield element


def cli_output(text, color: bool = False, stype: str = "", output_enabled: bool = True) -> None:
    """Print text like the legacy ``center.cliout`` helper."""
    if not output_enabled:
        return
    if color and len(text) > 0:
        text = " ".join(text.split())
        console = Console(width=len(text))
        console.print(
            Syntax(text.strip(), stype, word_wrap=True, indent_guides=True), end=""
        )
    else:
        print(text)


def debug_pair(text1, text, info_log: bool, output_enabled: bool = True, pretty_printer=None) -> None:
    """Print a labelled debug value like legacy ``center.x``."""
    if not (info_log and output_enabled):
        return
    pretty_printer = pretty_printer or pprint.PrettyPrinter(indent=4)
    if type(text) is str:
        print(str(text1) + ": " + text)
    else:
        print(str(text1) + ": ", end="")
        pretty_printer.pprint(text)


def debug_value(text, info_log: bool, output_enabled: bool = True, pretty_printer=None) -> None:
    """Print an unlabelled debug value like legacy ``center.alxp``."""
    if not (info_log and output_enabled):
        return
    pretty_printer = pretty_printer or pprint.PrettyPrinter(indent=4)
    if type(text) is str:
        print(text)
    else:
        pretty_printer.pprint(text)


def _doc_path(repo_root: Path, readme_filename: str) -> Path:
    return Path(repo_root) / "doc" / os.path.basename(readme_filename)


def reta_prompt_help_text(repo_root: Path, i18n) -> str:
    """Return the Markdown body printed by legacy ``retaPromptHilfe``."""
    place = _doc_path(Path(repo_root), i18n.readMeFileNames.retaPrompt)
    markdown_text = place.read_text(encoding="utf-8")
    start = markdown_text.find("+++", 2)
    markdown_text = re.sub(r"{#.*}", "", markdown_text)
    return markdown_text[start + 3 :]


def print_reta_prompt_help(repo_root: Path, i18n, console_factory=Console) -> None:
    """Render the retaPrompt help section to the console."""
    console = console_factory()
    console.print(Markdown(reta_prompt_help_text(repo_root, i18n)))


def reta_help_text(repo_root: Path, i18n) -> str:
    """Return the full reta help Markdown printed by legacy ``retaHilfe``."""
    place = _doc_path(Path(repo_root), i18n.readMeFileNames.reta)
    return place.read_text(encoding="utf-8")


def print_reta_help(repo_root: Path, i18n) -> None:
    """Print the reta help Markdown like legacy ``retaHilfe``."""
    print(reta_help_text(repo_root, i18n))


def get_text_wrap_things(max_len=None) -> tuple:
    """Return terminal wrapping helpers like legacy ``center.getTextWrapThings``."""
    if "Brython" not in sys.version.split():
        try:
            import html2text  # noqa: F401 - imported for historical side-effect/API parity
        except (ModuleNotFoundError, ImportError):
            html2text = None

        class _FallbackHyphenator:
            def wrap(self, text, width):
                if width <= 0:
                    return [text]
                return [text[i : i + width] for i in range(0, len(text), width)] or [""]

        try:
            import pyphen

            dic = pyphen.Pyphen(lang="de_DE")
        except (ModuleNotFoundError, ImportError):
            dic = _FallbackHyphenator()
        try:
            from textwrap2 import fill
        except (ModuleNotFoundError, ImportError):
            from textwrap import fill as _stdlib_fill

            def fill(text, width=70, **kwargs):
                kwargs.pop("use_hyphenator", None)
                return _stdlib_fill(text, width=width, **kwargs)

        h_de = None
        try:
            size = os.get_terminal_size()
            columns_amount, shell_rows_amount = size.columns, size.lines
        except OSError:
            try:
                if sys.stdin is not None and hasattr(sys.stdin, "isatty") and sys.stdin.isatty():
                    columns_amount, shell_rows_amount = os.popen("stty size", "r").read().split()
                else:
                    raise OSError
            except Exception:
                columns_amount, shell_rows_amount = "80", "80"
    else:  # pragma: no cover - Brython compatibility path
        html2text = None
        pyphen = None
        Hyphenator = None
        fill = None
        dic = None
        h_de = None
        columns_amount = "80"
        shell_rows_amount = "80"

    shell_width = int(columns_amount)
    return shell_width, h_de, dic, fill


class ConsoleIOMorphismBundle:
    """Activated architecture owner for center-level console and utility helpers."""

    def __init__(self, repo_root: str | Path | None = None, legacy_owner: str = "libs.center", activated_stage: int = 39):
        self.repo_root = Path(repo_root).resolve() if repo_root is not None else Path(__file__).resolve().parent.parent
        self.legacy_owner = legacy_owner
        self.activated_stage = activated_stage
        self.console_cls = Console
        self.markdown_cls = Markdown
        self.syntax_cls = Syntax

    def chunks(self, sequence: Sequence, size: int):
        return chunks(sequence, size)

    def unique_everseen(self, iterable: Iterable, key: Optional[Callable] = None, ordered_set_factory=None):
        return unique_everseen(iterable, key=key, ordered_set_factory=ordered_set_factory)

    def cliout(self, text, color: bool = False, stype: str = "", output_enabled: bool = True) -> None:
        return cli_output(text, color=color, stype=stype, output_enabled=output_enabled)

    def debug_pair(self, text1, text, info_log: bool, output_enabled: bool = True, pretty_printer=None) -> None:
        return debug_pair(text1, text, info_log=info_log, output_enabled=output_enabled, pretty_printer=pretty_printer)

    def debug_value(self, text, info_log: bool, output_enabled: bool = True, pretty_printer=None) -> None:
        return debug_value(text, info_log=info_log, output_enabled=output_enabled, pretty_printer=pretty_printer)

    def reta_prompt_help_text(self, i18n) -> str:
        return reta_prompt_help_text(self.repo_root, i18n)

    def print_reta_prompt_help(self, i18n) -> None:
        return print_reta_prompt_help(self.repo_root, i18n)

    def reta_help_text(self, i18n) -> str:
        return reta_help_text(self.repo_root, i18n)

    def print_reta_help(self, i18n) -> None:
        return print_reta_help(self.repo_root, i18n)

    def text_wrap_runtime(self, max_len=None) -> tuple:
        return get_text_wrap_things(max_len=max_len)

    def default_ordered_dict_type(self):
        return DefaultOrderedDict

    def snapshot(self) -> dict:
        return {
            "class": "ConsoleIOMorphismBundle",
            "stage": self.activated_stage,
            "legacy_owner": self.legacy_owner,
            "capsule": "OutputRenderingCapsule",
            "secondary_capsule": "InputPromptCapsule",
            "category": "ActivatedConsoleIOCategory",
            "functor": "ConsoleIOActivationFunctor",
            "natural_transformation": "CenterConsoleIOToArchitectureTransformation",
            "repo_root": str(self.repo_root),
            "morphisms": [
                "reta_prompt_help_text",
                "print_reta_prompt_help",
                "reta_help_text",
                "print_reta_help",
                "get_text_wrap_things",
                "cli_output",
                "debug_pair",
                "debug_value",
                "chunks",
                "unique_everseen",
                "DefaultOrderedDict",
            ],
            "compatibility_names": [
                "retaPromptHilfe",
                "retaHilfe",
                "getTextWrapThings",
                "cliout",
                "x",
                "alxp",
                "chunks",
                "unique_everseen",
                "DefaultOrderedDict",
            ],
            "observable_invariant": "center console/help/utility wrappers and ConsoleIOMorphismBundle expose the same visible output and finite-section helper results",
        }


def bootstrap_console_io_morphisms(repo_root: str | Path | None = None) -> ConsoleIOMorphismBundle:
    return ConsoleIOMorphismBundle(repo_root=repo_root)

File: retaRust/crates/reta_architecture/src/console_io.rs
//! Console/help/wrapping morphisms transcompiled from
//! `python_arch_reference/reta_architecture/console_io.py`.
//!
//! The Rust layer keeps these operations pure where possible.  Printing is
//! represented as rendered text decisions; callers may perform the actual side
//! effect at the facade boundary.

use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct DefaultOrderedDictSnapshot {
    pub class: String,
    pub default_factory: Option<String>,
    pub keys: Vec<String>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ConsoleIOSnapshot {
    pub class: String,
    pub stage: u32,
    pub legacy_owner: String,
    pub capsule: String,
    pub secondary_capsule: String,
    pub category: String,
    pub functor: String,
    pub natural_transformation: String,
    pub repo_root: String,
    pub morphisms: Vec<String>,
    pub compatibility_names: Vec<String>,
    pub observable_invariant: String,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct TextWrapRuntimeSnapshot {
    pub shell_width: usize,
    pub shell_rows_amount: usize,
    pub hyphenator: String,
    pub fill: String,
}

pub fn chunks<T: Clone>(sequence: &[T], size: usize) -> Vec<Vec<T>> {
    if size == 0 {
        return vec![sequence.to_vec()];
    }
    sequence.chunks(size).map(|chunk| chunk.to_vec()).collect()
}

pub fn unique_everseen<T, F>(iterable: impl IntoIterator<Item = T>, key: Option<F>) -> Vec<T>
where
    T: Clone + Ord,
    F: Fn(&T) -> String,
{
    let mut seen = BTreeSet::new();
    let mut out = Vec::new();
    for element in iterable {
        let marker = key
            .as_ref()
            .map(|func| func(&element))
            .unwrap_or_else(|| format_marker(&element));
        if seen.insert(marker) {
            out.push(element);
        }
    }
    out
}

fn format_marker<T: Ord>(value: &T) -> String {
    // The Python implementation stores the value itself in a set.  Rust keeps a
    // deterministic marker without requiring `Hash` on every caller type.
    format!("{:p}", value)
}

pub fn unique_strings_everseen(iterable: impl IntoIterator<Item = String>) -> Vec<String> {
    let mut seen = BTreeSet::new();
    let mut out = Vec::new();
    for element in iterable {
        if seen.insert(element.clone()) {
            out.push(element);
        }
    }
    out
}

pub fn cli_output_text(
    text: impl AsRef<str>,
    color: bool,
    stype: impl AsRef<str>,
    output_enabled: bool,
) -> Option<String> {
    if !output_enabled {
        return None;
    }
    let text = text.as_ref();
    if color && !text.is_empty() {
        Some(format!(
            "<syntax type=\"{}\">{}</syntax>",
            stype.as_ref(),
            text.split_whitespace().collect::<Vec<_>>().join(" ")
        ))
    } else {
        Some(text.to_string())
    }
}

pub fn debug_pair_text(
    text1: impl AsRef<str>,
    text: impl ToString,
    info_log: bool,
    output_enabled: bool,
) -> Option<String> {
    (info_log && output_enabled).then(|| format!("{}: {}", text1.as_ref(), text.to_string()))
}

pub fn debug_value_text(
    text: impl ToString,
    info_log: bool,
    output_enabled: bool,
) -> Option<String> {
    (info_log && output_enabled).then(|| text.to_string())
}

pub fn doc_path(repo_root: impl AsRef<Path>, readme_filename: impl AsRef<str>) -> PathBuf {
    let mut path = repo_root.as_ref().to_path_buf();
    path.push("doc");
    let basename = Path::new(readme_filename.as_ref())
        .file_name()
        .and_then(|value| value.to_str())
        .unwrap_or(readme_filename.as_ref());
    path.push(basename);
    path
}

pub fn strip_markdown_anchors(text: &str) -> String {
    let mut out = String::new();
    let mut rest = text;
    while let Some(start) = rest.find("{#") {
        out.push_str(&rest[..start]);
        if let Some(end) = rest[start..].find('}') {
            rest = &rest[start + end + 1..];
        } else {
            rest = &rest[start..];
            break;
        }
    }
    out.push_str(rest);
    out
}

pub fn reta_prompt_help_text_from_markdown(markdown_text: &str) -> String {
    let without_anchors = strip_markdown_anchors(markdown_text);
    match without_anchors
        .get(2..)
        .and_then(|tail| tail.find("+++").map(|idx| idx + 2))
    {
        Some(start) => without_anchors[start + 3..].to_string(),
        None => without_anchors,
    }
}

pub fn get_text_wrap_things(max_len: Option<usize>) -> TextWrapRuntimeSnapshot {
    let shell_width = max_len.filter(|value| *value > 0).unwrap_or_else(|| {
        std::env::var("COLUMNS")
            .ok()
            .and_then(|v| v.parse().ok())
            .unwrap_or(80)
    });
    let shell_rows_amount = std::env::var("LINES")
        .ok()
        .and_then(|v| v.parse().ok())
        .unwrap_or(80);
    TextWrapRuntimeSnapshot {
        shell_width,
        shell_rows_amount,
        hyphenator: "rust-fallback".to_string(),
        fill: "std-wrap".to_string(),
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct ConsoleIOMorphismBundle {
    pub repo_root: String,
    pub legacy_owner: String,
    pub activated_stage: u32,
}

impl ConsoleIOMorphismBundle {
    pub fn chunks<T: Clone>(&self, sequence: &[T], size: usize) -> Vec<Vec<T>> {
        chunks(sequence, size)
    }

    pub fn unique_strings_everseen(
        &self,
        iterable: impl IntoIterator<Item = String>,
    ) -> Vec<String> {
        unique_strings_everseen(iterable)
    }

    pub fn cliout(
        &self,
        text: impl AsRef<str>,
        color: bool,
        stype: impl AsRef<str>,
        output_enabled: bool,
    ) -> Option<String> {
        cli_output_text(text, color, stype, output_enabled)
    }

    pub fn debug_pair(
        &self,
        text1: impl AsRef<str>,
        text: impl ToString,
        info_log: bool,
        output_enabled: bool,
    ) -> Option<String> {
        debug_pair_text(text1, text, info_log, output_enabled)
    }

    pub fn debug_value(
        &self,
        text: impl ToString,
        info_log: bool,
        output_enabled: bool,
    ) -> Option<String> {
        debug_value_text(text, info_log, output_enabled)
    }

    pub fn doc_path(&self, readme_filename: impl AsRef<str>) -> PathBuf {
        doc_path(&self.repo_root, readme_filename)
    }

    pub fn reta_prompt_help_text_from_markdown(&self, markdown_text: &str) -> String {
        reta_prompt_help_text_from_markdown(markdown_text)
    }

    pub fn text_wrap_runtime(&self, max_len: Option<usize>) -> TextWrapRuntimeSnapshot {
        get_text_wrap_things(max_len)
    }

    pub fn default_ordered_dict_snapshot(&self) -> DefaultOrderedDictSnapshot {
        DefaultOrderedDictSnapshot {
            class: "OrderedDefaultDict".to_string(),
            default_factory: None,
            keys: Vec::new(),
        }
    }

    pub fn snapshot(&self) -> ConsoleIOSnapshot {
        ConsoleIOSnapshot {
            class: "ConsoleIOMorphismBundle".to_string(),
            stage: self.activated_stage,
            legacy_owner: self.legacy_owner.clone(),
            capsule: "OutputRenderingCapsule".to_string(),
            secondary_capsule: "InputPromptCapsule".to_string(),
            category: "ActivatedConsoleIOCategory".to_string(),
            functor: "ConsoleIOActivationFunctor".to_string(),
            natural_transformation: "CenterConsoleIOToArchitectureTransformation".to_string(),
            repo_root: self.repo_root.clone(),
            morphisms: vec![
                "reta_prompt_help_text".to_string(),
                "print_reta_prompt_help".to_string(),
                "reta_help_text".to_string(),
                "print_reta_help".to_string(),
                "get_text_wrap_things".to_string(),
                "cli_output".to_string(),
                "debug_pair".to_string(),
                "debug_value".to_string(),
                "chunks".to_string(),
                "unique_everseen".to_string(),
                "DefaultOrderedDict".to_string(),
            ],
            compatibility_names: vec![
                "retaPromptHilfe".to_string(),
                "retaHilfe".to_string(),
                "getTextWrapThings".to_string(),
                "cliout".to_string(),
                "x".to_string(),
                "alxp".to_string(),
                "chunks".to_string(),
                "unique_everseen".to_string(),
                "DefaultOrderedDict".to_string(),
            ],
            observable_invariant: "center console/help/utility wrappers and ConsoleIOMorphismBundle expose the same visible output and finite-section helper results".to_string(),
        }
    }
}

pub fn bootstrap_console_io_morphisms(repo_root: Option<String>) -> ConsoleIOMorphismBundle {
    ConsoleIOMorphismBundle {
        repo_root: repo_root.unwrap_or_else(|| ".".to_string()),
        legacy_owner: "libs.center".to_string(),
        activated_stage: 39,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn chunks_preserve_order() {
        assert_eq!(
            chunks(&[1, 2, 3, 4, 5], 2),
            vec![vec![1, 2], vec![3, 4], vec![5]]
        );
    }

    #[test]
    fn prompt_help_strips_anchors_after_frontmatter() {
        let result = reta_prompt_help_text_from_markdown("xx+++\n# A {#anchor}\nbody");
        assert!(result.contains("# A"));
        assert!(!result.contains("{#anchor}"));
    }
}
